chore: Update README.md (#67)
kubeflow/sdk: README.md

Kubeflow projects. It provides simple, consistent APIs across the Kubeflow ecosystem, enabling users to focus on building
ML applications rather than managing complex infrastrutcure.

### Kubeflow SDK Benefits

- **Unified Experience**: Single SDK to interact with multiple Kubeflow projects through consistent Python APIs
- **Simplified AI Workflows**: Abstract away Kubernetes complexity, allowing AI practitioners to work in familiar Python environments
- **Seamless Integration**: Designed to work together with all Kubeflow projects for end-to-end ML pipelines
- **Local Development**: First-class support for local development requiring only `pip` installation

<div style="text-align: center;">
  <img
    src="https://raw.githubusercontent.com/kubeflow/sdk/main/docs/images/persona_diagram.svg"
    width="600"
    title="Kubeflow SDK Personas"
    alt="Kubeflow SDK Personas"
  />
</div>

## Get Started

### Install Kubeflow SDK

```bash
pip install git+https://github.com/kubeflow/sdk.git@main
```
<!-- TODO(kramaranya): update before release pip install -U kubeflow -->

> ### Run your first distributed training job

Reviewer: ```suggestion
### Run your first PyTorch distributed job
```

- **Slack**: Join our [#kubeflow-ml-experience](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels) Slack channel
- **Meetings**: Attend the [Kubeflow SDK and ML Experience](https://bit.ly/kf-ml-experience) bi-weekly meetings
- **GitHub**: Discussions, issues and contributions at [kubeflow/sdk](https://github.com/kubeflow/sdk)

### Contributing

Kubeflow SDK is a community project and is still under active development. We welcome contributions! Please see our
[CONTRIBUTING Guide](https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md) for details.

## Documentation

<!-- TODO(kramaranya): add kubeflow sdk docs -->
- **[Design Document](https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit)**: Kubeflow SDK design proposal
- **[Component Guides](https://www.kubeflow.org/docs/components/)**: Individual component documentation
- **[DeepWiki](https://deepwiki.com/kubeflow/sdk)**: AI-powered repository documentation

## ✨ Contributors

We couldn't have done it without these incredible people:

<a href="https://github.com/kubeflow/sdk/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=kubeflow/sdk" />

> </a>

Reviewer: @kramaranya Do you know if PyPI is going to render it as well ?

Author: Yep! It works for https://pypi.org/project/llama-stack/ and https://pypi.org/project/feast/

# Kubeflow SDK

> <!-- TODO(kramaranya): update when release [![PyPI version](https://img.shields.io/pypi/v/kubeflow?color=%2334D058&label=pypi%20package)](https://pypi.org/project/kubeflow/) -->

Author: Yep, that makes sense!

[![Join Slack](https://img.shields.io/badge/Join_Slack-blue?logo=slack)](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels)
[![Coverage Status](https://coveralls.io/repos/github/kubeflow/sdk/badge.svg?branch=main)](https://coveralls.io/github/kubeflow/sdk?branch=main)

> [![License](https://img.shields.io/badge/License-Apache%202.0-blue)](https://github.com/kubeflow/sdk/blob/main/LICENSE)

Author: Thank you! I'll add it
However, why shouldn't we keep the License badge though if we'll publish it to PyPi?

> [![License](https://img.shields.io/badge/License-Apache%202.0-blue)](https://github.com/kubeflow/sdk/blob/main/LICENSE)

Reviewer: I just feel that it is redundant since PyPI already renders licence in the Meta section: https://pypi.org/project/kubeflow-trainer-api/#description
Same as GitHub.

> [![License](https://img.shields.io/badge/License-Apache%202.0-blue)](https://github.com/kubeflow/sdk/blob/main/LICENSE)

Reviewer: @kramaranya Did you push the commit to remove licence badge ?

[![Ask DeepWiki](https://deepwiki.com/badge.svg)](https://deepwiki.com/kubeflow/sdk)
<!-- TODO(kramaranya): update when release [![Python Supported Versions](https://img.shields.io/pypi/pyversions/kubeflow.svg?color=%2334D058)](https://pypi.org/project/kubeflow/) -->

## Overview

Kubeflow SDK is a unified Python SDK that streamlines the user experience for AI Practitioners to interact with various
Kubeflow projects. It provides simple, consistent APIs across the Kubeflow ecosystem, enabling users to focus on building
ML applications rather than managing complex infrastrutcure.

> **Note**: The Kubeflow SDK is in active development. Currently, the SDK provides full support for **Kubeflow

> Trainer** with other components planned for future releases.

Reviewer 2: Can we replace it with the Github style alerts? Maybe it would be better, just FYI.

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

### Kubeflow SDK Benefits

> - **Unified Experience**: Single SDK to interact with multiple Kubeflow components through consistent Python APIs

Reviewer: ```suggestion
- **Unified Experience**: Single SDK to interact with multiple Kubeflow projects through consistent Python APIs
```

Author: Updated! Shall we change https://www.kubeflow.org/docs/components/ to https://www.kubeflow.org/docs/projects/ btw? 🤔

Reviewer: We had discussion before with @thesuperzapper: https://github.com/kubeflow/website/issues/3982
It might be tricky to maintain redirects for each page if we change URL from `/components` to `/projects`.

Author: I see, thank you

> - **Simplified ML Workflows**: Abstract away Kubernetes complexity, allowing ML practitioners to work in familiar Python environments

Reviewer: I wouldn't overuse ML workflows term. 

```suggestion
- **Simplified AI Workloads**: Abstract away Kubernetes complexity, allowing AI practitioners to work in familiar Python environments
```

Author: updated

- **Seamless Integration**: Designed to work together with all Kubeflow projects for end-to-end ML pipelines
- **Local Development**: First-class support for local development requiring only `pip` installation

> 

Reviewer: Do we see any value to include persona diagram here ?
https://www.kubeflow.org/docs/components/trainer/overview/#who-is-this-for

Author: Since this diagram is specific to Trainer, I think that makes sense to include it for the first release and then create a more generic version for future releases as we add more KF projects? WDYT @andreyvelich @astefanutti @Electronic-Waste ?

Reviewer 3: @kramaranya that makes sense to me.

Author: I've added the diagram, let me know if there's anything I should change

Reviewer: Alternatively, we can re-use diagram from the SDK proposal: https://github.com/kubeflow/community/blob/master/proposals/819-kubeflow-sdk/kubeflow-sdk.drawio.svg

We might need to have better diagram in the future where we say that Kubeflow SDK is the primary interface for AI Practitioners to interact with Kubeflow projects APIs.

## Installation

Install the Kubeflow SDK using pip:

```bash

> pip install git+https://github.com/kubeflow/sdk.git@main

Reviewer: @kramaranya Did you push that change in this PR ?

```

> 

Reviewer: Maybe it worths to add simple getting started example, like in LangGraph: https://pypi.org/project/langgraph/

```py
from kubeflow.trainer import TrainerClient, CustomTrainer

def get_torch_dist():
    import os
    import torch
    import torch.distributed as dist

    dist.init_process_group(backend="gloo")
    print(f"PyTorch Distributed Environment")
    print(f"WORLD_SIZE: {dist.get_world_size()}")
    print(f"RANK: {dist.get_rank()}")
    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")

# Create the TrainJob.
job_id = TrainerClient().train(
    runtime=TrainerClient().get_runtime("torch-distributed"),
    trainer=CustomTrainer(
        func=get_torch_dist,
        num_nodes=3,
        resources_per_node={
            "cpu": 2,
        },
    ),
)

# Wait for TrainJob to complete.
TrainerClient().wait_for_job_status(job_id)

# Print TrainJob logs.
print(TrainerClient().get_job_logs(name=job_id, node_rank=0)["node-0"])
```

WDYT @kramaranya ?

Author: Yeah, I've been thinking about this too! I will add that

> ## Supported Components

Reviewer: ```suggestion
## Supported Kubeflow Projects
```

Author: Updated

| Component                   | Status | Description                                                    |
|-----------------------------|--------|----------------------------------------------------------------|
| **Kubeflow Trainer**        | ✅ **Available** | Train and fine-tune models with various frameworks   |
| **Kubeflow Katib**          | 🚧 Planned | Hyperparameter optimization                                |
| **Kubeflow Pipelines**      | 🚧 Planned | Build, run, and track ML workflows                         |
| **Kubeflow Spark Operator** | 🚧 Planned | Run Apache Spark jobs on Kubernetes                        |
| **Kubeflow Model Registry** | 🚧 Planned | Manage model artifacts, versions and ML artifacts metadata |
| **Kubeflow Feast**          | 🚧 Planned | Feature store for ML feature management and serving        |

> | **Kubeflow KServe**         | 🚧 Planned | Model deployment and serving infrastructure                |

Reviewer: Can we keep only Katib, Pipelines, and Registry here for now ?
We should have discussions on long-term approach for supported projects.

## Community

### Getting Involved

- **Slack**: Join our [#kubeflow-trainer](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels) Slack channel
  <!-- TODO(kramaranya): update to #ml-experience once we kick off -->
- **Meetings**: Attend the [AutoML and Training Working Group](https://bit.ly/2PWVCkV) bi-weekly meetings

>   <!-- TODO(kramaranya): update with kubeflow sdk meetings once we kick off -->

Reviewer: We can change it to Kubeflow SDK call: https://docs.google.com/document/d/1jH2WAX2ePxOfI4JuiVK9nPlesDMiyg67xzLwhpR7wTQ/edit?tab=t.0

- **GitHub**: Discussions, issues and contributions at [kubeflow/sdk](https://github.com/kubeflow/sdk)

### Contributing

Kubeflow SDK is a community project and is still under active development. We welcome contributions! Please see our

> [CONTRIBUTING Guide](CONTRIBUTING.md) for details.

Reviewer: Since this README is going to publish to PyPI, we might need to use complete URL for the docs.
```suggestion
[CONTRIBUTING Guide](https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md) for details.
```

Author: Good point, updated!

## Documentation

<!-- TODO(kramaranya): add kubeflow sdk docs -->
- **[Design Document](https://docs.google.com/document/d/1rX7ELAHRb_lvh0Y7BK1HBYAbA0zi9enB0F_358ZC58w/edit)**: Kubeflow SDK design proposal
- **[Component Guides](https://www.kubeflow.org/docs/components/)**: Individual component documentation
- **[DeepWiki](https://deepwiki.com/kubeflow/sdk)**: AI-powered repository documentation

## ✨ Contributors

We couldn't have done it without these incredible people:

<a href="https://github.com/kubeflow/sdk/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=kubeflow/sdk" />
</a>

## License

> Licensed under the [Apache License 2.0](LICENSE).

Reviewer: I don't think that we need it since Licence file presents in root.

Author: Makes sense, updated
